File: BiWeekly_Sprint/Sprint_25_10_15/AVL.cpp
#include <iostream>
#include <cstdlib>
#include <cstring>
#include <vector>
#include <algorithm>
#include <sstream>
#include <set>
#include <climits>
#include <cstdio>
#include <string>
using namespace std;
class AVL{
private:
	struct Node{
		int data;
		Node *left,*right;
		int height;
	};

	typedef struct Node * AVLNODE;
	AVLNODE root;
	
	AVLNODE NewNode(int data){
		Node *node = new Node();
		node->data = data;
		node->left = node->right = NULL;
		node->height = 1;
		return node;
	}
	int GetHeight(AVLNODE head){
		if(head == NULL)
			return 0;
		return max(head->left == NULL ? 0 : head->left->height,head->right==NULL ? 0 : head->right->height)+1;
	}
	int GetBalance(AVLNODE head){
		return GetHeight(head->left)-GetHeight(head->right);
	}

	AVLNODE LeftRotate(AVLNODE head,AVLNODE right){
		AVLNODE rightLeft = right->left;
		right->left = head;
		head->right = rightLeft;
		return right;
	}

	AVLNODE RightRotate(AVLNODE head,AVLNODE left){
		AVLNODE leftRight = left->right;
		left->right = head;
		head->left= leftRight;
		return left;		
	}
	void InsertHelper(AVLNODE &head,int data){
		if(head==NULL)
			head = NewNode(data);
		else if(data<head->data)
			InsertHelper(head->left,data);
		else
			InsertHelper(head->right,data);

		head->height = max(GetHeight(head->left),GetHeight(head->right))+1;

		if(GetBalance(head)> 1 && GetBalance(head->left) > 0)
			head = RightRotate(head,head->left);
		if(GetBalance(head)>1 && GetBalance(head->left) < 0 ){
			head->left = LeftRotate(head->left,head->left->right);
			head = RightRotate(head,head->left);
		}
		if(GetBalance(head) < -1 && GetBalance(head->right) < 0)
			head = LeftRotate(head,head->right);
		if(GetBalance(head) < -1 && GetBalance(head->right) > 0){
			head->right = RightRotate(head->right,head->right->left);
			head = LeftRotate(head,head->right);
		}
	}

	void DeleteHelper(AVLNODE &head,int data){
		if(head==NULL)
			return;
		else if(data<head->data)
			DeleteHelper(head->left,data);
		else if(data>head->data)
			DeleteHelper(head->right,data);
		else{
			if(head->left==NULL && head->right==NULL){
				delete head;
				head = NULL;
			}
			else if(head->left!=NULL && head->right == NULL){
				AVLNODE temp = head;
				head=head->left; delete temp;
			}
			else if(head->left==NULL && head->right!=NULL){
				AVLNODE temp = head;
				head=head->right; delete temp;
			}else{
				AVLNODE temp = head->right;
				while(temp->left!=NULL)
					temp = temp->left;
				swap(temp->data,head->data);
				DeleteHelper(head->right,data);
			}
		}

		if(head == NULL)
			return;

		head->height = max(GetHeight(head->left),GetHeight(head->right))+1;

		if(GetBalance(head)> 1 && GetBalance(head->left) > 0)
			head = RightRotate(head,head->left);
		if(GetBalance(head)>1 && GetBalance(head->left) < 0 ){
			head->left = LeftRotate(head->left,head->left->right);
			head = RightRotate(head,head->left);
		}
		if(GetBalance(head) < -1 && GetBalance(head->right) < 0)
			head = LeftRotate(head,head->right);
		if(GetBalance(head) < -1 && GetBalance(head->right) > 0){
			head->right = RightRotate(head->right,head->right->left);
			head = LeftRotate(head,head->right);
		}
	}
	void Inorder(AVLNODE temp){
		if(temp==NULL)
			return;
		Inorder(temp->left);
		cout<<temp->data<<" ";
		Inorder(temp->right);
	}
public:
	AVL(){
		root = NULL;
	}
	void Insert(int data){
		InsertHelper(root,data);
	}
	void Delete(int data){
		DeleteHelper(root,data);
	}
	void Display(){
		Inorder(root);
	}
};
int main(){
	std::ios_base::sync_with_stdio(false);
	freopen("1.txt","r",stdin);
	freopen("2.txt","w",stdout);
	int a[]={3,6,5,1,8,9,2,4,7,10};
	AVL avl;
	for(int i=0;i<10;i++)
		avl.Insert(a[i]);
	avl.Display();

	for(int i=9;i>=0;i--)
		avl.Delete(a[i]);
	for(int i=0;i<10;i++)
		avl.Insert(a[i]);
	avl.Display();

	for(int i=0;i<10;i++)
		avl.Delete(a[i]);
	avl.Display();
}

File: BiWeekly_Sprint/Sprint_4_10_15/AVL.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Text;
class AVL {
	class Node {
		public int data;
		public Node left, right;
		public int height;
	}

	private Node head;

	private Node NewNode(int data) {
		return new Node {data = data, height = 1};
	}
	private void InsertHelper(ref Node root, int data) {
		if (root ==  null)
			root = NewNode(data);
		else if (data < root.data)
			InsertHelper(ref root.left, data);
		else
			InsertHelper(ref root.right, data);

		root.height = Math.Max(Height(root.left), Height(root.right)) + 1;

		if (BalanceFactor(root) < -1)
		{
			if (BalanceFactor(root.right) < 0) {
				root = LeftRotate(root, root.right);
			}
			if (BalanceFactor(root.right) > 0) {
				root.right = RightRotate(root.right, root.right.left);
				root = LeftRotate(root, root.right);
			}
		}
		if (BalanceFactor(root) > 1) {
			if (BalanceFactor(root.left) > 0) {
				root = RightRotate(root, root.left);
			}
			if (BalanceFactor(root.left) < 0) {
				root.left = LeftRotate(root.left, root.left.right);
				root = RightRotate(root, root.left);
			}
		}
	}

	public void Insert(int data) {
		InsertHelper(ref head, data);
	}

	private void DisplayHelper(Node root) {
		if (root != null) {
			DisplayHelper(root.left);
			Console.Write(root.data + " ");
			DisplayHelper(root.right);
		}
	}
	public void Display() {
		DisplayHelper(head);
	}

	private void Swap<T>(ref T a, ref T b) {
		T c = a;
		a = b;
		b = c;
	}
	private void DeleteHelper(ref Node root, int data) {
		if (root == null)
			return;

		if (data < root.data)
			DeleteHelper(ref root.left, data);
		else if (data > root.data)
			DeleteHelper(ref root.right, data);
		else {
			if (root.left == null && root.right == null) {
				root = null;
			} else if (root.left != null && root.right == null) {
				var temp = root;
				root = root.left;
				temp = null;
			} else if (root.left == null && root.right != null) {
				var temp = root;
				root = root.right;
				temp = null;
			} else {
				var temp = root.right;
				while (temp.left != null)
					temp = temp.left;

				Swap(ref temp.data, ref root.data);
				DeleteHelper(ref root.right, data);
			}

		}

		if (root == null)
			return;
		root.height = Math.Max(Height(root.left), Height(root.right)) + 1;

		if (BalanceFactor(root) < -1)
		{
			if (BalanceFactor(root.right) < 0) {
				root = LeftRotate(root, root.right);
			}
			if (BalanceFactor(root.right) > 0) {
				root.right = RightRotate(root.right, root.right.left);
				root = LeftRotate(root, root.right);
			}
		}
		if (BalanceFactor(root) > 1) {
			if (BalanceFactor(root.left) > 0) {
				root = RightRotate(root, root.left);
			}
			if (BalanceFactor(root.left) < 0) {
				root.left = LeftRotate(root.left, root.left.right);
				root = RightRotate(root, root.left);
			}
		}

	}

	public void Delete(int data) {
		DeleteHelper(ref head, data);
	}

	private int Height(Node root) {
		if (root == null)
			return 0;

		return (root.left == null ? 0 : root.left.height) + (root.right == null ? 0 : root.right.height);
	}

	private int BalanceFactor(Node root) {
		if (root == null)
			return 0;

		return Height(root.left) - Height(root.right);
	}

	private Node LeftRotate(Node curr, Node right) {
		Node rightLeft = right.left;
		right.left = curr;
		curr.right = rightLeft;
		return rightLeft;
	}

	private Node RightRotate(Node curr, Node left) {
		Node leftRight = left.right;
		left.right = curr;
		curr.left = leftRight;
		return left;
	}
	static void Main() {
		Console.WriteLine("I love you");
		int [] arr = {4, 6, 1, 8, 3, 2, 9, 0, 5, 7};
		var avl = new AVL();
		foreach (var v in arr)
			avl.Insert(v);

		avl.Display();
		Console.Write("\n");

		foreach (var v in arr) {
			//Console.Write(v);
			avl.Delete(v);
			avl.Display();
			Console.Write("\n");
		}

	}
}